persistent_stream.go

```diff
@@ -0,0 +1,348 @@
+	// process request
+	pid, err := peer.IDFromBytes(req.GetCallUnary().Peer)
+	if err != nil {
+		return errorUnaryCall(callID, err)
+	}
+
+	// open stream to remote
+	remoteStream, err := d.host.NewStream(
+		ctx,
+		pid,
+		protocol.ID(*req.GetCallUnary().Proto),
+	)
+	if err != nil {
+		return errorUnaryCall(callID, err)
+	}
+	defer remoteStream.Close()
+
+	select {
+	case response := <-exchangeMessages(ctx, remoteStream, req):
+		return response
+
+	case <-ctx.Done():
+		return okCancelled(callID)
+	}
+}
+
+func exchangeMessages(ctx context.Context, s network.Stream, req *pb.PersistentConnectionRequest) <-chan *pb.PersistentConnectionResponse {
+	callID, _ := uuid.FromBytes(req.CallId)
+	rc := make(chan *pb.PersistentConnectionResponse)
+
+	go func() {
+		defer close(rc)
+
+		// write request to remote
+		if err := ggio.NewDelimitedWriter(s).WriteMsg(req); ctx.Err() != nil {
+			return
+		} else if err != nil {
+			rc <- errorUnaryCall(callID, err)
+			return
+		}
+
+		// await response from remote
+		remoteResp := &pb.PersistentConnectionRequest{}
+		if err := ggio.NewDelimitedReader(s, network.MessageSizeMax).ReadMsg(remoteResp); ctx.Err() != nil {
+			return
+		} else if err != nil {
+			rc <- errorUnaryCall(callID, err)
+			return
+		}
+
+		// convert response to a persistent channel response
+		resp := okUnaryCallResponse(callID)
+		resp.Message = &pb.PersistentConnectionResponse_CallUnaryResponse{
+			CallUnaryResponse: remoteResp.GetUnaryResponse(),
+		}
+
+		// avoid blocking rc channel on cancelled context
+		select {
+		case rc <- resp:
+			return
+
+		case <-ctx.Done():
+			return
+		}
+	}()
+
+	return rc
+}
+
+// awaitReadFail writers to a semaphor channel if the given io.Reader fails to
+// read before the context was cancelled
+func awaitReadFail(ctx context.Context, r io.Reader) <-chan struct{} {
+	semaphor := make(chan struct{})
+
+	go func() {
+		defer close(semaphor)
+
+		buff := make([]byte, 1)
+		if _, err := r.Read(buff); err != nil {
+			select {
+			case semaphor <- struct{}{}:
+			case <-ctx.Done():
+			}
+		}
+	}()
+
+	return semaphor
+}
+
+// getPersistentStreamHandler returns a lib-p2p stream handler tied to a
+// given persistent client stream
+func (d *Daemon) getPersistentStreamHandler(cw ggio.Writer) network.StreamHandler {
+	return func(s network.Stream) {
+		defer s.Close()
+
+		// read request from remote peer
+		req := &pb.PersistentConnectionRequest{}
+		if err := ggio.NewDelimitedReader(s, network.MessageSizeMax).ReadMsg(req); err != nil {
+			log.Debugw("failed to read proto from incoming p2p stream", "error", err)
+			return
+		}
+
+		// now the peer field points to the caller
+		req.GetCallUnary().Peer = []byte(s.Conn().RemotePeer())
+
+		callID, err := uuid.FromBytes(req.CallId)
+		if err != nil {
+			log.Debugw("bad call id in p2p handler", "error", err)
+			return
+		}
+
+		// create response channel
+		rc := make(chan *pb.PersistentConnectionRequest)
+		d.responseWaiters.Store(callID, rc)
```

Comment:
```suggestion
		responseChan := make(chan *pb.PersistentConnectionRequest)
		d.responseWaiters.Store(callID, responseChan)
```

I suggest to avoid the obvious comments and the comments that can be replaced with a proper variable names. The reason is that comments require time to maintain and can easily become unsynchronized with the actual code, while the variable names usually can't.